core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -135,207 +146,285 @@ class PartitionStateMachine(controller: KafkaController, stateChangeLogger: Stat
+    val adjustedLeaderAndIsrs = partitionsWithLeaders.map { case (partition, leaderAndIsrOpt, recipients) => partition -> leaderAndIsrOpt.get }.toMap
+    val (successfulUpdates, updatesToRetry, failedUpdates) = zkUtils.updateLeaderAndIsr(adjustedLeaderAndIsrs, controllerContext.epoch)
+    successfulUpdates.foreach { case (partition, leaderAndIsr) =>
+      val replicas = controllerContext.partitionReplicaAssignment(partition)
+      val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(leaderAndIsr, controllerContext.epoch)
+      controllerContext.partitionLeadershipInfo.put(partition, leaderIsrAndControllerEpoch)
+      controllerBrokerRequestBatch.addLeaderAndIsrRequestForBrokers(recipientsPerPartition(partition), partition.topic,
+        partition.partition, leaderIsrAndControllerEpoch, replicas, isNew = false)
+    }
+    (successfulUpdates.keys.toSeq, updatesToRetry, failedElections.toMap ++ failedUpdates.toMap)
   }
 
-  /**
-   * Invoked on the OfflinePartition,OnlinePartition->OnlinePartition state change.
-   * It invokes the leader election API to elect a leader for the input offline partition
-   * @param topic               The topic of the offline partition
-   * @param partition           The offline partition
-   * @param leaderSelector      Specific leader selector (e.g., offline/reassigned/etc.)
-   */
-  def electLeaderForPartition(topic: String, partition: Int, leaderSelector: PartitionLeaderSelector) {
-    val topicAndPartition = TopicAndPartition(topic, partition)
-    val stateChangeLog = stateChangeLogger.withControllerEpoch(controller.epoch)
-    // handle leader election for the partitions whose leader is no longer alive
-    stateChangeLog.trace(s"Started leader election for partition $topicAndPartition")
-    try {
-      var zookeeperPathUpdateSucceeded: Boolean = false
-      var newLeaderAndIsr: LeaderAndIsr = null
-      var replicasForThisPartition: Seq[Int] = Seq.empty[Int]
-      while(!zookeeperPathUpdateSucceeded) {
-        val currentLeaderIsrAndEpoch = getLeaderIsrAndEpochOrThrowException(topic, partition)
-        val currentLeaderAndIsr = currentLeaderIsrAndEpoch.leaderAndIsr
-        val controllerEpoch = currentLeaderIsrAndEpoch.controllerEpoch
-        if (controllerEpoch > controller.epoch) {
-          val failMsg = s"Aborted leader election for partition $topicAndPartition since the LeaderAndIsr path was " +
-            s"already written by another controller. This probably means that the current controller $controllerId went " +
-            s"through a soft failure and another controller was elected with epoch $controllerEpoch."
-          stateChangeLog.error(failMsg)
-          throw new StateChangeFailedException(stateChangeLog.messageWithPrefix(failMsg))
+  private def leaderForOffline(leaderIsrAndControllerEpochs: Seq[(TopicAndPartition, LeaderIsrAndControllerEpoch)]) = {
```

Comment:
Could we explicitly define the return type in this and the following private method? Otherwise, it's a bit hard to follow the logic.